momepy/preprocessing.py

```diff
@@ -740,3 +748,240 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+        rab_plus['center_pt'] = gdf[gdf.index == gdf.index_right].geometry.centroid
+    
+    elif center_type == 'mean':
+        # mean geometry
+        ls_xy = [g.exterior.coords.xy for g in rab_plus.geometry] #extracting the points
+        
+        ls_xy = [point for polygon in multipolygon for point in polygon.exterior.coords[:-1]]
+        
+        mean_pts = [geometry.Point(np.mean(xy[0]),np.mean(xy[1])) for xy in ls_xy]
+        rab_plus['center_pt'] = gpd.GeoSeries( data = mean_pts, 
+                                              index = rab_plus.index, 
+                                              crs = edges.crs)
+    
+    # centerpoint of minimum_bounding_circle
+    # minimun_bounding_circle() not available in Shapely 1.8.2 but only in 'latest'
+    # --> https://shapely.readthedocs.io/en/latest/constructive.html
+    # current environment has 1.8.2
+    
+    return rab_plus
+
+def _coins_filtering_many_incoming(incoming_many, angle_threshold=0):
+    """
+    Used only for the cases when more than one incoming line touches the
+    roundabout.
+    """
+    # figuring out which one needs to be extended and retain attributes
+    coins_filter_result = []
+    # For each new connection, evaluate COINS and selecet the group from which the new line belongs
+    for g, x in incoming_many.groupby('line_wkt'):
+        gs = gpd.GeoSeries( pd.concat([x.geometry, x.line]), crs= incoming_many.crs )
+        gdf = gpd.GeoDataFrame(geometry = gs)
+        gdf.drop_duplicates(inplace=True)
+
+        coins = COINS(gdf, angle_threshold=angle_threshold)
+        stroke_gdf = coins.stroke_gdf()
+
+        orig_geom_join = stroke_gdf.sjoin(gpd.GeoDataFrame(geometry = x.line), predicate= 'covers' )
+        orig_geom = gpd.GeoSeries([orig_geom_join.geometry.iloc[0]], crs= incoming_many.crs)
+        gs2 = gpd.GeoDataFrame(geometry = orig_geom)
+
+        gs1 = gpd.GeoSeries(x.geometry, crs= incoming_many.crs )
+        gs1 = gpd.GeoDataFrame(geometry = gs1)
+
+        #select the the line that's covered by the joined line returned by COINS
+        result_idx = gs1.sjoin(gs2 , predicate = 'covered_by').index
+        coins_filter_result.extend(result_idx)
+    
+    incoming_many_reduced = incoming_many.loc[coins_filter_result]
+    return incoming_many_reduced
+
+def _selecting_incoming_lines (rab_plus, edges, angle_threshold=0):
+    """Selecting only the lines that are touching but not covered by
+    the ``rab_plus``.
+    If more than one LineString is incoming to ``rab_plus``, COINS algortthm
+    is used to select the line to be extended further.
+    """
+    # Feels a bit combersome ... Ideally there would be a DISJOINT predicate
+    incoming = edges.sjoin(rab_plus , predicate = 'touches')
+    incoming.rename(columns ={'index_right':'index_rab_plus'}, inplace = True )
+    idx_drop =  incoming.sjoin(rab_plus, predicate = 'covered_by').index
+    incoming.drop(idx_drop, axis=0, inplace =True)
```

Comment:
You could create your own query on `.sindex` here and make the filter before doing the actual join.

Reply:
Not sure I quite follow what you mean by this. 
I will keep the comment open to review later